API review changes

HaoK opened the pull request:
Rename IHostingFactory -> IHostingBuilder and move Use methods from IHostingEngine
Rename HostingFactory -> WebHostBuilder
Move HostingEngine to internal

Fixes https://github.com/aspnet/Hosting/issues/236

cc @davidfowl @lodejard @Tratcher

davidfowl commented on src/Microsoft.AspNet.Hosting/Internal/RootHostingServiceCollectionInitializer.cs:
```diff
@@ -43,7 +41,7 @@ public IServiceCollection Build()
             services.AddInstance(_loggerFactory);
 
-            services.AddTransient<IHostingFactory, HostingFactory>();
+            services.AddTransient<IHostingBuilder, WebHostBuilder>();
```
This shouldn't be a service. It's something you new up with no state.

davidfowl commented:
Some extra notes that weren't in the bug
- WebHost shouldn't exist anymore
- WebHostBuilder is *not* a service and there doesn't need to be an interface
- IConfiguration is optional to WebHostBuilder and should not be an argument to Build.
- WebHostBuilder has 2 ctors -> WebHostBuilder(IServiceProvider), WebHostBuilder(IServiceProvider, IConfiguration)
- Program.cs shouldn't have any code that reads configuration. The WebHostBuilder uses defaults from configuration when the constructor that takes IConfiguration is used.

HaoK commented:
@davidfowl Can we expose UseConfiguration instead of ctor overload?

HaoK commented:
Updated PR, nuked WebHost

HaoK commented:
Split IStartupLoader methods into FindStartupType/LoadMethods which enables `UseStartup<T>` also moved environment off interface methods, service can get it from IHostingEnvironment

davidfowl commented:
> @davidfowl Can we expose UseConfiguration instead of ctor overload?

We decided against that since config is something that needs to be ready super early. There were other reasons as well /cc @glennc @lodejard

davidfowl commented on src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs:
```diff
@@ -0,0 +1,164 @@
+
+        public HostingEngine(IServiceCollection appServices, IStartupLoader startupLoader, IConfiguration config)
+        {
+            _config = config ?? new Configuration();
```
This should never be null

davidfowl commented on src/Microsoft.AspNet.Hosting/Program.cs:
```diff
@@ -23,31 +24,37 @@ public Program(IServiceProvider serviceProvider)
-                engine.UseServer(server);
+                builder.UseServer(server);
             }
             var startup = config.Get("app");
```
This code should be gone. Passing the configuration to the WebHostBuilder makes this logic obsolete as the builder should use the config system passed in as a sign to set these default.

HaoK replied:
So we want to bake "server" and "app" in to the builder as default config keys for UseServer and UseStartup?  Should we pick new new keys?

davidfowl replied:
might be the time to pick new keys /cc @lodejard. server and app are terrible. Lets keep new app and server as a fallback. Maybe something like:

``
var server = config.Get("Hosting:Server") ?? config.Get("server");
var app = config.Get("Hosting:Application") ?? config.Get("app");
``

davidfowl replied:
Also do the same for Hosting:Environment?

HaoK replied:
Okay that looks much better, I can do the same thing with the environment name: "ASP_ENV" or "Hosting:Environment"

davidfowl replied:
Although I realized today that on *nix env variables can't contain `';'`.

HaoK replied:
So we want config to win right?  Where config is now going to override anything set problematically before hand?  Or do we want config to be the defaults applied in the ctor?  and UseStartup will win?

HaoK replied:
Today it looks like config Environment wins over UseEnvironment (but I'm not sure that's what we want either...)

davidfowl commented on src/Microsoft.AspNet.Hosting/WebHostBuilder.cs:
```diff
@@ -0,0 +1,184 @@
+        {
+            _hostingEnvironment = new HostingEnvironment();
+            _loggerFactory = new LoggerFactory();
+            _fallbackServices = fallbackServices ?? CallContextServiceLocator.Locator.ServiceProvider;
```
Drop support for CallContextServiceLocator.Locator.ServiceProvider, we're removing this across the board. It can stay in test host for now but all mainline APIs should remove it.

HaoK replied:
So [NotNull] for IServiceProvider then?

davidfowl replied:
Yep

davidfowl commented on src/Microsoft.AspNet.TestHost/TestServer.cs:
```diff
@@ -48,51 +48,51 @@ public static TestServer Create(Action<IApplicationBuilder> configureApp, Action
-        {
-            return Create<TStartup>(fallbackServices: null, config: null, configureApp: null, configureServices: null);
-        }
+        //public static TestServer Create<TStartup>() where TStartup : class
```
Dead code?

HaoK replied:
yep nuked and cleaned up TestServer

davidfowl commented on src/Microsoft.AspNet.Hosting/WebHostBuilder.cs:
```diff
@@ -0,0 +1,184 @@
+
+        public WebHostBuilder() : this(null) { }
+
+        public WebHostBuilder(IServiceProvider fallbackServices)
```
Rename fallbackServices to services

HaoK replied:
So today after nuking the default ctor(), we are calling GetRequiredService on the ISP, so things go kerboom if you try to pass in any arbitrary ISP...  What behavior do we want here?  The tests need to either pass in the old CallContextServiceLocator, or we can fix the import to allow this?

HaoK replied:
Actually I guess the hosting engine tests have to pass the CallContextSeviceLocator since it relies on the services there...

HaoK commented:
Addressed PR feedback

davidfowl commented on test/Microsoft.AspNet.TestHost.Tests/TestServerTests.cs:
```diff
@@ -56,53 +56,53 @@ public void ThrowsIfNoApplicationEnvironmentIsRegisteredWithTheProvider()
-            string result = await server.CreateClient().GetStringAsync("/path");
-            Assert.Equal("AppPath:" + appPath, result);
-        }
+        //[Fact]
```
What's the deal with these tests?

HaoK replied:
I used to have support on TestServer for overriding the ApplicationBasePath/ApplicationName, not sure we still need this with the config changes to allow setting the base path directly?  I ripped out the support for now and MVC functionals pass without this, so I think I can just nuke this once and for all

HaoK replied:
Nuked the AppPathName tests, reenabled the Create<T> tests for TestServer

HaoK commented:
Pushed config server/application key logic into WebHostBuilder

HaoK commented:
ping @davidfowl 

@glennc nag Fowler if you want to play with these changes soon

HaoK commented:
Rebased and fixed tests to accommodate new RequestIdentifier changes

HaoK commented:
234bbf82f2329e1ce2092962264ba5d4567d115b
